### PR TITLE
[WIP] 16 create ec2 source

### PR DESCRIPTION
**Pull request by dylanratcliffe**

### CODE REVIEW

File: sources/ec2/az.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		input = ec2.DescribeAvailabilityZonesInput{
+			ZoneNames: []string{
+				query,
+			},
+		}
+	case sdp.RequestMethod_LIST:
+		input = ec2.DescribeAvailabilityZonesInput{}
+	default:
+		return nil, errors.New("unsupported method")
+	}
```
Comment (DavidS-om) [inline sources/ec2/az.go:28]: This feels like it'll become repetitive very quickly. maybe have an `AvailabilityZone{Get,List,Search}InputMapper` instead?

File: sources/ec2/az.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	return &input, nil
+}
+
+// AvailabilityZoneOutputMapper Maps API output to items
+func AvailabilityZoneOutputMapper(scope string, output *ec2.DescribeAvailabilityZonesOutput) ([]*sdp.Item, error) {
+	if output == nil {
+		return nil, errors.New("empty output")
+	}
+
+	items := make([]*sdp.Item, len(output.AvailabilityZones))
```
Comment (DavidS-om) [inline sources/ec2/az.go:39]: nice optimization

File: sources/ec2/ec2_source.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	// DescribeFunc for a given set of scope, query and method
+	InputMapper func(scope, query string, method sdp.RequestMethod) (Input, error)
+
+	// A function that returns the input object will be passed to DescribeFunc
+	// for a given set of scope, query and method. This also includes a
+	// maxResults param which should be fed to the API to mean "page size", and
+	// a nextToken string pointer: If this is the first page this will be nil,
+	// however for subsequest pages this will be populated with the appropriate
+	// token
+	InputMapperPaginated func(scope, query string, method sdp.RequestMethod, maxResults *int32, nextToken *string) (Input, error)
```
Comment (DavidS-om) [inline sources/ec2/ec2_source.go:40]: should this return/wrap [Paginators](https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#using-paginators) where possible? Looking at the paginator implementation, there's more happening than meets the eye, e.g. this `HasMorePages` method has more conditions than our impl does:

```go
func (p *DescribeInstancesPaginator) HasMorePages() bool {
	return p.firstPage || (p.nextToken != nil && len(*p.nextToken) != 0)
}
```

**Review by DavidS-om — COMMENTED** [review]
review comments to discuss

**Comment by dylanratcliffe** [comment]
@DavidS-om I've addressed all the comments so I'm going to close this and press on to reduce spam. It's looking waaaaaay nicer now

**Review by DavidS-om — APPROVED** [review]
good stuff

**Pull request closed without merging**